#3156 PostgreSQL support

Patrykb0802 opened the pull request.

Patrykb0802 committed:
#3156 PostgreSQL support

Limraj committed:
#3144 Fixed data point name characters are incorrectly escaped:
- Fixed fields: name/xid for all Data Source;
- Fixed OPC UA fields: nodeName, identifier, attributes;

Limraj commented on src/org/scada_lts/dao/migration/postgresql/V1_1__PostgresViewsHierarchy.java:
```diff
@@ -0,0 +1,280 @@
+import org.springframework.jdbc.core.JdbcTemplate;
+
+
+public class V1_1__PostgresViewsHierarchy extends BaseJavaMigration {
```
The classes are in their packages, there is no need to add the word postgres to the migration class name. The difference between migration classes and other classes is that other classes are/will be used in the sources, then it is better for the class itself to have information about what implementation it concerns, whereas migration classes should not be used in the sources, they have one purpose - for migration using the flyway library.

Limraj commented on src/org/scada_lts/dao/migration/postgresql/V1_1__PostgresViewsHierarchy.java:
```diff
@@ -0,0 +1,280 @@
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.scada_lts.dao.migration.postgresql;
```
Sufficient and consistent package name: org.scada_lts.dao.migration.postgres . If it is possible to shorten the name while retaining the same meaning, then we shorten it, and shorter names are more readable. Also, we use the word postgres in the code.

Limraj commented on src/org/scada_lts/dao/migration/postgresql/V1_1__PostgresViewsHierarchy.java:
```diff
@@ -0,0 +1,280 @@
+						+ "$$ LANGUAGE plpgsql;";
+
+		String pSelect =
+				"CREATE OR REPLACE FUNCTION prc_views_hierarchy_select() "
```
If there was a procedure in MySQL, there must also be a procedure in Postgres.

Limraj commented on src/org/scada_lts/dao/DAO.java:
```diff
@@ -169,5 +175,9 @@ public boolean isTest() {
 	}
-	
+
+	public static DatabaseAccess.DatabaseType getType() {
```
This should be an instance method

Limraj commented on src/com/serotonin/mango/db/PostgreSQLAccess.java:
```diff
@@ -73,46 +72,68 @@ protected String getDriverClassName() {
+            if (flyway == null) {
+                flyway = Flyway.configure()
+                        .dataSource(getDataSource())
+                        .locations("org.scada_lts.dao.migration.postgresql")
```
org.scada_lts.dao.migration.postgres

Limraj commented on src/org/scada_lts/cache/DataSourcePointsCache.java:
```diff
@@ -69,8 +73,15 @@ public boolean isCacheEnabled() {
+			dao = new PostgresDataPointDAO();
+		} else {
+			dao = new DataPointDAO();
+		}
```
IDataPointDAO dao = ApplicationBeans.getBean("dataPointDAO", IDataPointDAO.class);

Limraj commented on src/org/scada_lts/cache/UpdateDataSourcesPoints.java:
```diff
@@ -20,8 +24,14 @@ public class UpdateDataSourcesPoints implements StatefulJob{
+                dao = new PostgresDataPointDAO();
+            } else {
+                dao = new DataPointDAO();
+            }
```
IDataPointDAO dao = ApplicationBeans.getBean("dataPointDAO", IDataPointDAO.class);

Limraj commented on WebContent/WEB-INF/applicationContext.xml:
```diff
@@ -79,8 +79,6 @@
   <bean id="usersProfileDAO" class="org.scada_lts.dao.UsersProfileDAO" />
   <bean id="dataPointUserDAO" class="org.scada_lts.dao.DataPointUserDAO"/>
   <bean id="viewDAO" class="org.scada_lts.dao.ViewDAO"/>
-  <bean id="watchListDAO" class="org.scada_lts.dao.watchlist.WatchListDAO"/>
```
```
  <bean id="watchListDaoFactory" class="org.scada_lts.factory.WatchListDaoFactory">
    <constructor-arg ref="databaseAccess"/>
  </bean>

  <bean id="watchListDAO" factory-bean="watchListDaoFactory" factory-method="newInstance"/>
```

Limraj commented on WebContent/WEB-INF/applicationContext.xml:
```diff
@@ -79,8 +79,6 @@
   <bean id="dataPointUserDAO" class="org.scada_lts.dao.DataPointUserDAO"/>
   <bean id="viewDAO" class="org.scada_lts.dao.ViewDAO"/>
-  <bean id="watchListDAO" class="org.scada_lts.dao.watchlist.WatchListDAO"/>
-  <bean id="dataSourceDAO" class="org.scada_lts.dao.DataSourceDAO"/>
```
```
<bean id="dataSourceDaoFactory" class="org.scada_lts.factory.DataSourceDaoFactory">
   <constructor-arg ref="databaseAccess"/>
 </bean>

 <bean id="dataSourceDAO" factory-bean="dataSourceDaoFactory" factory-method="newInstance"/>
```

Limraj commented on WebContent/WEB-INF/applicationContext.xml:
```diff
@@ -168,10 +166,8 @@
   <!-- -->
 
   <!-- Point Hierarchy -->
-  <bean id="dataPointDAO" class="org.scada_lts.dao.DataPointDAO"/>
```
We leave the old lines of code in place, we modify and add an additional bean definition. Notice that the file has comments that tell us what the bean is for.

```
  <bean id="dataPointDaoFactory" class="org.scada_lts.factory.DataPointDaoFactory">
    <constructor-arg ref="databaseAccess"/>
  </bean>

  <bean id="dataPointDAO" factory-bean="dataPointDaoFactory" factory-method="newInstance"/>
```

Limraj commented on src/org/scada_lts/service/pointhierarchy/PointHierarchyService.java:
```diff
@@ -244,19 +245,19 @@ public List<PointHierarchyConsistency> checkPointHierarchyConsistency() {
 
 	public PointHierarchyNode getPointHierarchyRoot(User user)  {
-		return PointHierarchyUtils.getPointHierarchyRoot(user, hierarchyDAO, dataPointDAO);
+		return PointHierarchyUtils.getPointHierarchyRoot(user, hierarchyDAO, (DataPointDAO) dataPointDAO);
```
If the PostgresDataPointDAO implementation is injected into the class, then when trying to execute this method and cast this object to the DataPointDAO class, we will get an exception: ClassCastException.

DataPointDAO is a specific implementation, it has no connection with PostgresDataPointDAO, although both classes implement IDataPointDAO, it is like saying that a bus is a passenger car, although both are vehicles, have gearboxes, wheels, etc., they have some common elements, the general structure may be similar, the principle of operation, but they are fundamentally different, hence the interface is used to solve such problems.
Where the DataPointDAO class is used, it must be replaced with the IDataPointDAO interface. How to do it cleverly? With the help of the refactoring option you can change the interface name to IIDataPointDAO, then in the applicationContext.xml for the DataPointDAO class we change its name to IDataPointDAO and using the refactoring option, we did the same for the class itself and now without refactoring we go back to the old name DataPointDAO and IDataPointDAO and in this way we limit the number of specific changes, now those places where the DataPointDAO class was used are associated with the IDataPointDAO interface, and DataPointDAO with a specific implementation. Of course this still requires corrections, there will be places `new IDataPointDAO()` which will indicate a compilation error.

Limraj committed:
#2162 Avoid double input in data point properties edition:
- Removed unsupported unit 'kilometers';

Patrykb0802 committed:
#3156 PostgreSQL support

Limraj committed:
#2746 Added Assignee function for events:
- Fixed long time login;

Limraj committed:
#2746 Added Assignee function for events:
- Corrected HighestAlarmLevelService;

Limraj committed:
#3088 Fixed StackOverflowError for Cyclic Dependency in Meta Data Source:
- Fixed running long time;

Limraj committed:
#3088 Fixed StackOverflowError for Cyclic Dependency in Meta Data Source:
- Corrected CollectMetaDataPointFromContextAction;

Limraj committed:
#3088 Fixed StackOverflowError for Cyclic Dependency in Meta Data Source:
- Corrected logging in CollectMetaDataPointFromContextAction;
- Reduced depth for CollectMetaDataPointFromContextAction;

Patrykb0802 committed:
#3156 PostgreSQL support

Limraj committed:
#2755 Signaling the Data Source that has a problem:
- Fixed Modbus for refuse connection;
- Fixed active events for None alarm level;

Limraj committed:
#2985 Prevent XSS for REST API by escape String content:
- Unescaped text field in Link Component;

Limraj committed:
#3088 Fixed StackOverflowError for Cyclic Dependency in Meta Data Source:
- Corrected CollectMetaDataPointFromContextAction;
- Corrected method StartStopDataPointsUtils.getSequenceMetaDataPoints;

Limraj committed:
#3147 Fixed Modbus IP unsupported data type:
- Fixed Modbus IP for slaveId = 0;
- Removed deprecated methods from ModbusSerialDataSource RT class: getPort, verifyPort;

Patrykb0802 committed:
#3156 PostgreSQL support

Limraj committed:
#3169 Corrected v2.8.0:
- Fixed the Modbus Serial form; (editModbusSerial.jsp)
- Reduced the number of calculations for the unreliable flag for a point;
- Corrected intervalTime in common.js;
- Added properties in env.properties: scadalts.datapoint.unreliable.depth, scadalts.datapoint.unreliable.execute-in-pool-if-tasks-exceeds, set 5 for property: recursive-pool-executor.keep-alive-time;
- Added 'org.scada_lts.recursive' Logger in log4j2.xml;
- Added tests: SetDataPointUnreliableUtilsTest, ResetDataPointUnreliableUtilsTest, GetRunningMetaDataPointsTest, CyclicGetRunningMetaDataPointsTest;

Limraj committed:
#3169 Corrected v2.8.0:
- Fixed MQTT restoring the connection after it has been lost;

Patrykb0802 committed:
#3156 PostgreSQL support

Limraj committed:
#3169 Corrected v2.8.0:
- Reduced setUnreliableDataPoint: added method DataSourceRT.doSetUnreliableDataPoint ;
- Fixed BinaryValue - use trim;
- Added tests: BinaryValueParseBinaryTest, BinaryValueParseBinaryExceptionTest;
- Fixed cancel events after disable Event Handler;
- Set default updateEvent as UPDATE_EVENT_CONTEXT_CHANGE in Meta Data Source;

Limraj committed:
#3169 Corrected v2.8.0:
- Corrected Modbus validation, increased timeout range to 180 000 ms

Patrykb0802 committed:
Revert "#3169 Corrected v2.8.0:"

This reverts commit 2eb21c1c9ab41de3140d85c93d2b136709e590c9.

Patrykb0802 committed:
Revert "#3169 Corrected v2.8.0:"

This reverts commit 9de9a629190e962a82198f69ba72b828aa9f5aed.

Patrykb0802 committed:
Revert "#3169 Corrected v2.8.0:"

This reverts commit 3212105171a9667b5e809382fd7aae8d7d320956.

Patrykb0802 committed:
Revert "#3169 Corrected v2.8.0:"

This reverts commit d65e3993f861fc11b9b25a57564c06e09249350e.

Patrykb0802 committed:
Revert "#3147 Fixed Modbus IP unsupported data type:"

This reverts commit bbd701799456b4d247846eca79c4df612e079b55.

Patrykb0802 committed:
Revert "#2985 Prevent XSS for REST API by escape String content:"

This reverts commit 15fb10db69850818e362794545c6b66257f2ab02.

Patrykb0802 committed:
Revert "#3088 Fixed StackOverflowError for Cyclic Dependency in Meta Data Source:"

This reverts commit b70c0988c22814685c0d210de00761b36e896d37.

Patrykb0802 committed:
Revert "#2755 Signaling the Data Source that has a problem:"

This reverts commit 806016d6f3e66e388c0cd71c479399ef7d7045fc.

Patrykb0802 committed:
Revert "#3088 Fixed StackOverflowError for Cyclic Dependency in Meta Data Source:"

This reverts commit 41efd129b689c0cad7ca4e593b682124e1ca2f53.

Patrykb0802 committed:
Revert "#3088 Fixed StackOverflowError for Cyclic Dependency in Meta Data Source:"

This reverts commit 31715c20417ffe20389d45490e5478456978d0ad.

Patrykb0802 committed:
Revert "#3088 Fixed StackOverflowError for Cyclic Dependency in Meta Data Source:"

This reverts commit 896ed29f6183bb6fc83fdc964994369fed5ab884.

Patrykb0802 committed:
Revert "#2746 Added Assignee function for events:"

This reverts commit 8302eaaed86c7af6bf85ee31dac3df5fa2f60458.

Patrykb0802 committed:
Revert "#2746 Added Assignee function for events:"

This reverts commit 49b7dd9fdae8e4aa7d5afa417eb0409ad2d56916.

Patrykb0802 committed:
Revert "#2162 Avoid double input in data point properties edition:"

This reverts commit 8f764fa73f4d8938f9648b1f3a57fe5434d44d3a.

Patrykb0802 committed:
Revert "#3144 Fixed data point name characters are incorrectly escaped:"

This reverts commit b91fc90a2d88c16cea4bbec4d1cb34312bfe6e60.
